### PR TITLE
[REEF-1648] Allow transition from ALLOCATED to RUNNING in EvaluatorState

### PR DESCRIPTION
**Pull request by tcNickolas**
JIRA:
  [REEF-1648](https://issues.apache.org/jira/browse/REEF-1648)

Pull request:
  This closes #

### CODE REVIEW

**Comment by tcNickolas** [comment]
I've tested this change by running O.A.R.Examples.DriverRestart on an HDI cluster, and the job succeeded. Previously it got stuck because evaluators kept trying to connect to the driver, and driver kept rejecting them.

**Comment by tcNickolas** [comment]
Good AppVeyor build: https://ci.appveyor.com/project/tcNickolas/reef/build/315-REEF-1648

**Comment by tcNickolas** [comment]
Driver only sets evaluator to `SUBMITTED` if it actually submits task to it (which makes sense). 

On the first run of driver it submits a task to evaluator and switches it to `SUBMITTED`. After driver restart evaluator is initialized in `ALLOCATED` state, and driver doesn't set them to `SUBMITTED` again (because it doesn't submit task to old evaluator after restart). So evaluator contacts driver to report it's `RUNNING` and reconnect, and driver doesn't allow the transition.

**Comment by markusweimer** [comment]
> Driver only sets evaluator to `SUBMITTED` if it actually submits task to it (which makes sense).

Slight correction: It also sets it to `SUBMITTED` if only a root context was submitted to it.

**Comment by tcNickolas** [comment]
True. But after restart driver doesn't submit either task or context to an evaluator which already has something running. Driver needs to accept heartbeat from evaluator first to figure out that this evaluator is still alive before it can decide whether it needs to submit something to it.

**Comment by jwang98052** [comment]
When the driver is restarted, should the existing evaluator state is initialized to the previous state? It could be ALLOCATED or SUBMITTED depending on when the driver was down, right?

**Comment by tcNickolas** [comment]
How would driver after restart know what driver before restart did? We're only storing container allocations and deallocations, not every event of evaluator lifetime. We shouldn't store every event for performance reasons, since we're writing every change to DFS, and we haven't yet done performance testing on real scenarios to see how much overhead it introduces. And even if we wrote every event, it's possible that one of the writes will fail, which would lead us to exactly the same situation as we're seeing now.

I consider the check for allowed transitions as something to prohibit "backwards" transitions, like from completed to running, not to impose extra restrictions on "forward" transitions.

**Comment by jwang98052** [comment]
Let's put comments so that we know why we allow this transition. 

For more extensive level of driver fault tolerant, I believe there will be a lot of more scenarios to cover. Or it can be left to the driver.

**Comment by tcNickolas** [comment]
I've added a comment to describe the scenario in which such transition can occur.

File: lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorState.java [hunk 1]
```diff
@@ -122,6 +122,9 @@ public final boolean isLegalTransition(final EvaluatorState toState) {
     case ALLOCATED:
       switch(toState) {
       case SUBMITTED:
+      // direct transition to RUNNING can happen in case of driver restart,
```
Comment (markusweimer) [inline lang/java/reef-common/src/main/java/org/apache/reef/runtime/common/driver/evaluator/EvaluatorState.java]: Can we gate this with a check that the Driver is indeed being restarted?

**Comment by markusweimer** [comment]
OK, I will test and merge.

**Comment by markusweimer** [comment]
I merged this, but our github mirror seems to be behind.

**Comment by tcNickolas** [comment]
Weird, this change is not large enough to block our mirror. We should merge some other change on top of this one to unblock - very carefully :-) I'll close this for now

**Pull request closed without merging**